### ISSUE TITLE
Notify Harbormaster for non-differential builds

### ISSUE THREAD

**Issue by pzinovkin** [uber/phabricator-jenkins-plugin#108]
Now it's ignored:
> [phabricator:non-differential] Coverage result not found. Ignoring build.

It would be nice to have it for diffusion to show build results.

<img width="492" alt="screenshot 2015-11-02 00 00 54" src="https://cloud.githubusercontent.com/assets/440630/10871331/18d51964-80f5-11e5-86b9-743b3a98af6d.png">

**Comment by sectioneight** [comment]
Yup, out workflow at Uber doesn't utilize this but this seems like a generally useful and easy to implement feature.

**Comment by sectioneight** [comment]
Can you clarify how you're triggering these jenkins jobs? Do you have a herald rule on autoclose commits, or is Jenkins polling your SCM master branch?

**Comment by pzinovkin** [comment]
It's a herald rule for commits. It runs build plan which triggers job in jenkins. Job triggered with parameters `GIT_COMMIT=${buildable.commit}&PHID=${target.phid}`

**Comment by altendky** [comment]
I have the following herald rules setup on my test server.

![image](https://cloud.githubusercontent.com/assets/543719/11337085/39f7a28c-91b8-11e5-81c9-723dc21855e7.png)

![image](https://cloud.githubusercontent.com/assets/543719/11337076/2939b250-91b8-11e5-8e56-9682fdda54ff.png)

When I make a new commit it does start a build.

![image](https://cloud.githubusercontent.com/assets/543719/11337171/da343044-91b8-11e5-85b2-041e26c7a50c.png)
![image](https://cloud.githubusercontent.com/assets/543719/11337205/2fb274b8-91b9-11e5-8194-65506d0b5406.png)

And the Jenkins log does show the build:

```
Started by user Phabricator
Building in workspace /var/lib/jenkins/jobs/Test/workspace
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url http://p.fstab.net/diffusion/T/test.git # timeout=10
Fetching upstream changes from http://p.fstab.net/diffusion/T/test.git
 > git --version # timeout=10
using .gitcredentials to set credentials
 > git config --local credential.username jenkins # timeout=10
 > git config --local credential.helper store --file=/tmp/git8096642345063786895.credentials # timeout=10
 > git -c core.askpass=true fetch --tags --progress http://p.fstab.net/diffusion/T/test.git +refs/heads/*:refs/remotes/origin/*
 > git config --local --remove-section credential # timeout=10
 > git rev-parse refs/remotes/origin/master^{commit} # timeout=10
 > git rev-parse refs/remotes/origin/origin/master^{commit} # timeout=10
Checking out Revision bb711d40fdcaf2b809f8e0fce4bcfcf0b2a89d5f (refs/remotes/origin/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f bb711d40fdcaf2b809f8e0fce4bcfcf0b2a89d5f
 > git rev-list 41e245b1706ed2ee75830f7a39c2933944903fba # timeout=10
[phabricator:ignore-build] No differential ID found.
[workspace] $ /bin/sh -xe /tmp/hudson2902405382799744766.sh
+ make
mkdir -p results
echo 'Direct commit - 3' > results/test
Archiving artifacts
[phabricator:non-differential] Coverage result not found. Ignoring build.
Finished: SUCCESS
```

This is in contrast to a differential triggered build which does respond to the build request and close it.

```
Started by user Phabricator
Building in workspace /var/lib/jenkins/jobs/Test/workspace
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url http://p.fstab.net/diffusion/T/test.git # timeout=10
Fetching upstream changes from http://p.fstab.net/diffusion/T/test.git
 > git --version # timeout=10
using .gitcredentials to set credentials
 > git config --local credential.username jenkins # timeout=10
 > git config --local credential.helper store --file=/tmp/git5505462600999545426.credentials # timeout=10
 > git -c core.askpass=true fetch --tags --progress http://p.fstab.net/diffusion/T/test.git +refs/heads/*:refs/remotes/origin/*
 > git config --local --remove-section credential # timeout=10
 > git rev-parse refs/remotes/origin/master^{commit} # timeout=10
 > git rev-parse refs/remotes/origin/origin/master^{commit} # timeout=10
Checking out Revision 4473ea55aa39b66c45c9071e270d3fe4677963b6 (refs/remotes/origin/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 4473ea55aa39b66c45c9071e270d3fe4677963b6
 > git rev-list 4473ea55aa39b66c45c9071e270d3fe4677963b6 # timeout=10
[phabricator:conduit] Fetching differential from Conduit API
[workspace] $ git reset --hard 4473ea55aa39b66c45c9071e270d3fe4677963b6
HEAD is now at 4473ea5 Testing differential
[workspace] $ git clean -fd -f
Removing results/
[workspace] $ git submodule update --init --recursive
[workspace] $ /var/lib/phabricator/arcanist/bin/arc patch --nobranch --diff 10 --nocommit ********
Checking patch makefile...
Applied patch makefile cleanly.
 OKAY  Successfully applied patch.
[workspace] $ /bin/sh -xe /tmp/hudson7389794144036638857.sh
+ make
mkdir -p results
echo 'Test 2 - D2-1' > results/test
Archiving artifacts
[phabricator:process-build-result] No line coverage found, skipping...
[phabricator:process-build-result] No unit results available.
[phabricator:process-build-result] No coverage provider available.
[phabricator:harbormaster] Sending Harbormaster BUILD_URL via PHID: PHID-HMBT-2ianqbrzd4ero25fxqik
[phabricator:process-build-result] Sending build result to Harbormaster with PHID PHID-HMBT-2ianqbrzd4ero25fxqik, success: true
[phabricator:comment-file] no files found by path: '.phabricator-comment'
Finished: SUCCESS
```

**Comment by sectioneight** [comment]
Thanks for the info! I think I have a plan to modify the logic.

The part where it gets tricky is that a lot of the code is built around the
idea that we can submit a comment (for the differential), but Conduit
doesn't contain methods to comment on Diffusion. So all you'll get back is
a build status. Is that sufficient for your needs?

**Comment by altendky** [comment]
That would be a nice start.  I looked and agree that I do not see a conduit method for commenting on diffusion (or audit) though you certainly can comment on them...  interesting.

Except!  https://secure.phabricator.com/T6468
Not sure how comfortable you are with a hidden/deprecated function.

**Comment by pzinovkin** [comment]
Build status is good enough for me.

**Comment by sectioneight** [comment]
Cool. I'll try to add this this weekend.

**Comment by sectioneight** [comment]
I put up a branch here: https://github.com/uber/phabricator-jenkins-plugin/tree/non-differential-status-108

It doesn't have tests yet, but should implement the functionality desired.

Could one of you try installing it and see if it meets your needs?

To install a development version, pull down the repo and run `mvn package` which will generate a .hpi file in the `target/` directory which you can install on your jenkins under the plugin manager's "advanced" tab.

If this is the right approach, I'll add tests and get it reviewed.

**Comment by altendky** [comment]
![image](https://cloud.githubusercontent.com/assets/543719/11401886/de394732-9362-11e5-8ff4-e33db799d02a.png)

Looks like it's working.  I got myself confused and went through a bunch of other stuff that included having to install the Cobertura plugin for Jenkins to avoid exceptions.  This certainly may be unrelated to this particular change.

My main interest in the comments from Jenkins are to have a link from the Phabricator build to the Jenkins build.  Presently, the only way I see to do this is manually via the commit hash.  Is there another way I am missing?

As always, thanks for your work on this.

**Comment by sectioneight** [comment]
When you say "have a link from the Phabricator build to the Jenkins build" -- are you talking about this part?

![image](https://cloud.githubusercontent.com/assets/97087/11407306/d27f1512-9366-11e5-94b5-aae17cf931eb.png)

That is the Harbormaster "artifact" that it doesn't look like I implemented for non-differentials (d'oh).

If that's the only piece you're missing, I'll go ahead and add that and tests, get this reviewed and merged.

If there's something else that's missing, let me know.

**Comment by sectioneight** [comment]
... and if that is the feature you were missing,  I just pushed another untested commit to the branch that should accomplish this.

**Comment by altendky** [comment]
I just got home from work and that's where my test rig is.  I won't be back until next Wednesday.

If the image you show is of a hyperlink from within Phabricator to the specific URL of the corresponding Jenkins build, then yes, that's what I'm looking for.

I'm not sure if I want to or not but it crossed my mind to have the build results transmitted back to Phabricator as well, but that can certainly be left for another time if not already covered by your reference to artifacts...  if I even want that to happen.

Thanks for getting this implemented.  It certainly is a nice integration.

**Comment by sectioneight** [comment]
That image is from a differential. I'm not sure what the UI looks like on
the diffusion side. My test VM is also unavailable until I get back to work
next week. I'll see if I can get something running if I have some spare
time this weekend so I can stop using you as a guinea pig :)

**Comment by hach-que** [comment]
This functionality is critical for us as we drive both commit builds and release pipelines through Phabricator, and we need to be able to build arbitrary commits in Jenkins and have the result returned to a PHID (even if a Diff ID is not specified).

**Comment by sectioneight** [comment]
Sounds good. Could you try pulling down my branch linked above and see if
it meets your needs? It this is the right approach, I'll add tests and get
this merged.

**Comment by ChaitanyaPramod** [comment]
I'm looking forward to this very much, yay to more visibility of the build system into phabricator. I checked the branch commits and they seem good to me.

@sectioneight Is there anything else we can do to help close this?

**Comment by sectioneight** [comment]
No, I just need to revive my branch and write tests.

**Comment by Hunsu** [comment]
I'm too looking forward for this feature. What I'm interested in is to have a direct link from Phabricator to the build console.

**Comment by hach-que** [comment]
@sectioneight I meant to update this, but being a Windows shop we don't have the capacity to build Java plugins for Jenkins.  We'll only be able to test if this is working when we can install the plugin through the Jenkins interface.

**Comment by sectioneight** [comment]
Confirmed that the branch works, including the Jenkins link from the artifact.

Diffusion commit page:

![image](https://cloud.githubusercontent.com/assets/97087/12430669/b744e898-bea5-11e5-8d1d-185528ed6627.png)

Diffusion index page:

![image](https://cloud.githubusercontent.com/assets/97087/12430677/be0a51ea-bea5-11e5-837c-671f7bd5267e.png)

Going to write some tests and get this merged today. Thanks for the patience.

**Comment by sectioneight** [comment]
This is merged and released in 1.9.0, which I just uploaded to maven. You should see it in the your update manager within the next 12 hours.

**Comment by pzinovkin** [comment]
Hi there. Nice work. But it seems to skip uberalls reporting.
